### PR TITLE
bisect--helper: `bisect_clean_state` shell function in C

### PR DESCRIPTION
**Pull request by pranitbauva1997**
A first look at the bisect_clean_state() function.

There are no error messages if the file or directory isn't removed in the shell function. I wonder whether we could add those in the shell implementation and the C function?

### CODE REVIEW

File: builtin/bisect--helper.c [hunk 1]
```diff
@@ -89,12 +90,35 @@ int bisect_voc(const char *term){
 	return 0;
 }
 
+int bisect_clean_state(void)
+{
+	struct strbuf bisect_refs = STRBUF_INIT;
+
+	strbuf_addstr(&bisect_refs, ".git/refs/bisect/");
+	remove_dir_recursively(&bisect_refs, 0);
```
Comment (chriscool) [inline builtin/bisect--helper.c]: Please don't deal with ref stuff like that. See my comment in: https://github.com/pranitbauva1997/git/pull/8

Comment (chriscool) [inline builtin/bisect--helper.c]: And also what happens if refs are packed?

Comment (pranitbauva1997) [inline builtin/bisect--helper.c]: This is just a bare version. I read man pages and they have strictly warned about this. I am finding ways to remove those refs.

Comment (pranitbauva1997) [inline builtin/bisect--helper.c]: refs can be packed? I am quite unaware about this fact. Will search about this more.

Comment (chriscool) [inline builtin/bisect--helper.c]: Try `git help pack-refs`.

Comment (chriscool) [inline builtin/bisect--helper.c]: In general if you want to know how to properly replace a shell git command with some C functions, you can take a look at what C functions the command is calling.

To change refs the git-bisect.sh uses `git update-ref`, so you could find interesting things by looking at its implementation.

Comment (pranitbauva1997) [inline builtin/bisect--helper.c]: @chriscool Thanks! I will do it.

**Comment by pranitbauva1997** [comment]
It has been some time I am kind of stuck for quite some time now at a segmentation fault. I will try for one more day. If not possible then I will ask for your help.

After this I will be aiming for small function conversion like bisect_head(), is_expected_rev(), check_expected_rev(), bisect_write(). Hopefully if I am not stuck at a seg fault in these too, I can finish off these in this week.

**Comment by pranitbauva1997** [comment]
I tried a variety of things like first storing SHA1 in an array and then deleting them, directly deleting them, using strbuf for string handling, various conversions from hex back and forth. I haven't properly diagnosed the problem too. Just a small help, could you please point out how to use gdb for testing this kind of problem. I searched but they were mostly describing how to debug one source file programs. I am not at all comfortable with debugging with gdb. Thanks for you time!

**Comment by chriscool** [comment]
Here is an example of an unfinished gdb session to debug the bisect_clean_state() function:

```
$ gdb git 
GNU gdb (Ubuntu 7.7.1-0ubuntu5~14.04.2) 7.7.1
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from git...done.
(gdb) b bisect_clean_state
Breakpoint 1 at 0x41a06b: file builtin/bisect--helper.c, line 109.
(gdb) run bisect--helper --bisect-clean-state
Starting program: /home/utilisateur/git/git/git bisect--helper --bisect-clean-state
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, bisect_clean_state () at builtin/bisect--helper.c:109
warning: Source file is more recent than executable.
109             for_each_ref_in("refs/bisect/", remove_bisect_refs, NULL);
(gdb) l
104     }
105
106     int bisect_clean_state(void)
107     {
108
109             for_each_ref_in("refs/bisect/", remove_bisect_refs, NULL);
110
111
112             remove_path(".git/BISECT_EXPECTED_REV");
113             remove_path(".git/BISECT_ANCESTORS_OK");
(gdb) n
112             remove_path(".git/BISECT_EXPECTED_REV");
(gdb) n
113             remove_path(".git/BISECT_ANCESTORS_OK");
(gdb) n
114             remove_path(".git/BISECT_LOG");
(gdb) h
List of classes of commands:

aliases -- Aliases of other commands
breakpoints -- Making program stop at certain points
data -- Examining data
files -- Specifying and examining files
internals -- Maintenance commands
obscure -- Obscure features
running -- Running the program
stack -- Examining the stack
status -- Status inquiries
support -- Support facilities
tracepoints -- Tracing of program execution without stopping the program
user-defined -- User-defined commands

Type "help" followed by a class name for a list of commands in that class.
Type "help all" for the list of all commands.
Type "help" followed by command name for full documentation.
Type "apropos word" to search for commands related to "word".
Command name abbreviations are allowed if unambiguous.
(gdb) 
```

You can try to enter the same commands at the shell prompt "$ " and at the gdb prompt "(gdb)".
It should help you understand how to use gdb.

If there are things you don't understand above, please ask specific questions.

**Comment by pranitbauva1997** [comment]
@chriscool Thanks a lot! `run bisect--helper --bisect-clean-state` was the line which I couldn't figure out. I will get back to you as soon as I fix this error

**Comment by pranitbauva1997** [comment]
@chricool Thanks for your gdb help! I am able to figure out the reason for segfault. I am passing oid->hash (which is a SHA1 hex value) where I should pass something like commit id values (I don't really know what they are called). So now I have to see how to convert those hex values into "normal values" to pass them.

Setting break points at functions helped a lot! :+1:

**Comment by pranitbauva1997** [comment]
The segmentation fault is created when the previous test already executed "git bisect reset" and the the first thing the new test executes is "git bisect reset". I am digging more into this.

**Comment by pranitbauva1997** [comment]
@chriscool @larsxschneider I have still not fixed the bug. Just for some change I have now started working on is_expected_rev() and check_expected_rev(). I will get back on bisect_clean_state() after this

**Comment by pranitbauva1997** [comment]
@chriscool @larsxschneider I have identified the problem. It is in refs/files-backend.c in the function do_for_each_entry_in_dir() . Basically what is happening that the dir struct is having more entries than it is supposed to have and the remaining ones point to a memory location which does not exist at all. I am now trying to fix the error in refs code.

**Comment by pranitbauva1997** [comment]
@chriscool @larsxschneider Should I send an RFC to the mailing list and see responses of people who have worked on the refs backend ?

**Comment by chriscool** [comment]
If you think it is ready, you can always send an RFC patch series to the mailing list.
In this case from the patches I see above, it doesn't look like it is ready. So maybe you could push here an updated version first.

**Comment by pranitbauva1997** [comment]
@chriscool Sure!

**Comment by pranitbauva1997** [comment]
@chriscool @larsxschneider I have done the cleanups. Review it once before I send it to the mailing list.

File: builtin/bisect--helper.c [hunk 2]
```diff
@@ -79,18 +91,54 @@ int write_terms(const char *bad, const char *good)
 	strbuf_release(&content);
 	return (res < 0) ? -1 : 0;
 }
+
+int remove_bisect_refs(const char *refname, const struct object_id *oid,
```
Comment (chriscool) [inline builtin/bisect--helper.c]: It looks like this is removing only one ref, so you could remove the 's' at the end of the name.

File: builtin/bisect--helper.c [hunk 3]
```diff
@@ -79,18 +91,54 @@ int write_terms(const char *bad, const char *good)
...
+		       int flag, void *cb_data)
+{
+	struct strbuf temp = STRBUF_INIT;
+	strbuf_addf(&temp, "refs/bisect/%s", refname);
```
Comment (chriscool) [inline builtin/bisect--helper.c]: Maybe using xstrfmt() could save one line.

File: builtin/bisect--helper.c [hunk 4]
```diff
@@ -79,18 +91,54 @@ int write_terms(const char *bad, const char *good)
...
+	if (delete_ref(temp.buf, oid->hash, flag))
+		return error(_("couldn't delete the ref %s\n"), temp.buf);
+	strbuf_release(&temp);
+	return 0;
+}
+
+int bisect_clean_state(void)
+{
+
```
Comment (chriscool) [inline builtin/bisect--helper.c]: This blank line should be removed.

File: builtin/bisect--helper.c [hunk 5]
```diff
@@ -79,18 +91,54 @@ int write_terms(const char *bad, const char *good)
...
+	for_each_ref_in("refs/bisect/", remove_bisect_refs, NULL);
+
+
```
Comment (chriscool) [inline builtin/bisect--helper.c]: This blank line too.

**Comment by pranitbauva1997** [comment]
Till that time I would study about various plumbing commands used in git-bisect as they would be soon converted. And maybe I could find some help in ref handling.

File: builtin/bisect--helper.c [hunk 6]
```diff
@@ -109,6 +157,8 @@ int cmd_bisect__helper(int argc, const char **argv, const char *prefix)
 		if (argc != 2)
 			die(_("--write-terms requires two arguments"));
 		return write_terms(argv[0], argv[1]);
+	case BISECT_CLEAN_STATE:
+		return bisect_clean_state();
```
Comment (chriscool) [inline builtin/bisect--helper.c]: Maybe check 'argc' just before the above line.

File: builtin/bisect--helper.c [hunk 7]
```diff
@@ -79,18 +91,54 @@ int write_terms(const char *bad, const char *good)
...
+
+	remove_path(git_path_bisect_expected_rev());
+	remove_path(git_path_bisect_ancestors_ok());
+	remove_path(git_path_bisect_log());
+	remove_path(git_path_bisect_names());
+	remove_path(git_path_bisect_run());
+	remove_path(git_path_bisect_terms());
+	/* Cleanup head-name if it got left by an old version of git-bisect */
+	remove_path(git_path_head_name());
+	remove_path(git_path_bisect_head());
```
Comment (chriscool) [inline builtin/bisect--helper.c]: BISECT_HEAD is a ref, so it should be deleted using delete_ref().

File: builtin/bisect--helper.c (hunk 6 above)
Comment (pranitbauva1997) [inline builtin/bisect--helper.c]: Thanks for your comments. I will update and then send it!

**Comment by pranitbauva1997** [comment]
@chriscool @larsxschneider I have finished work on this. Though I believe the error handling could be improved. What are your thoughts? I am now starting with converting bisect_reset(). Thanks for all the help!

File: builtin/bisect--helper.c [hunk 8]
```diff
@@ -3,13 +3,26 @@
...
+static GIT_PATH_FUNC(git_path_bisect_start, "BISECT_START")
 
 static const char * const git_bisect_helper_usage[] = {
 	N_("git bisect--helper --next-all [--no-checkout]"),
 	N_("git bisect--helper --write-terms <bad_term> <good_term>"),
+	N_("git bisect--helper --bisect-clean-state"),
 	NULL
 };
 
+static struct string_list refs_for_removal;
```
Comment (chriscool) [inline builtin/bisect--helper.c]: Are you sure this struct needs to be static?
Couldn't you pass a pointer to the instance in the "void *cb_data" parameter below?

File: builtin/bisect--helper.c [hunk 9]
```diff
@@ -79,18 +92,52 @@ int write_terms(const char *bad, const char *good)
...
+	string_list_clear(&refs_for_removal, 0);
+	remove_path(git_path_bisect_expected_rev());
+	remove_path(git_path_bisect_ancestors_ok());
+	remove_path(git_path_bisect_log());
+	remove_path(git_path_bisect_names());
+	remove_path(git_path_bisect_run());
+	remove_path(git_path_bisect_terms());
+	/* Cleanup head-name if it got left by an old version of git-bisect */
+	remove_path(git_path_head_name());
+	result |= delete_ref("BISECT_HEAD", NULL, 0);
```
Comment (chriscool) [inline builtin/bisect--helper.c]: Maybe you could just add "BISECT_HEAD" to "refs_for_removal" string_list above.

**Comment by chriscool** [comment]
After taking care of the comments I just made, it would be nice if you could resend another version of this to the mailing list. But this can wait for a few days.
Otherwise ok for converting bisect_reset(). Thanks!

File: builtin/bisect--helper.c (hunk 8 above)
Comment (pranitbauva1997) [inline builtin/bisect--helper.c]: Yes I could. I will look more into this. Thanks!

File: builtin/bisect--helper.c (hunk 9 above)
Comment (pranitbauva1997) [inline builtin/bisect--helper.c]: Sure!

**Comment by pranitbauva1997** [comment]
I guess I will wait for a few days before sending this to the mailing list.

**Pull request closed without merging**